### DISCUSSION TITLE
How to add additional fields to contract definitions

### DISCUSSION THREAD

**Discussion by pablobiedma** [eclipse-edc/Connector#2472] [Q&A] [answered]
Dear all, I am a beginner using the connector and playing around with the 'Samples' tutorial. I am trying to add additional fields to a contract definition and I was wondering what the steps for doing this are.

Suppose we wanted to add a String field called `'test 1'`.
This is what I've tried so far:

Adding such a field to the `contractoffer.json` file
```
{
  "connectorId": "provider",
  "connectorAddress": "http://localhost:8282/api/v1/ids/data",
  "protocol": "ids-multipart",
  "test1": "test1",
...
```

Adding the field to the `ContractDefinition.java` class, creating a getter and a builder for it.

```
private String test1;
```

```
    @NotNull
    public String getTest1() {
        return test1;
    }
```

```
        public Builder test1(String test1) {
            entity.test1 = test1;
            return this;
        }
```

I did the same for the `ContractOffer.java` class

I also went to the `FileTransferExtension.java` class and added the field to the `registerContractDefinition` method: 

```
    private void registerContractDefinition(String uid) {
        var contractDefinition = ContractDefinition.Builder.newInstance()
                .id("1")
                .accessPolicyId(uid)
                .contractPolicyId(uid)
                .selectorExpression(AssetSelectorExpression.Builder.newInstance()
                        .whenEquals(Asset.PROPERTY_ID, "test-document")
                        .build())
                .validity(TimeUnit.HOURS.toSeconds(1))
                .test1("This string should be displayed")
                .build();

        contractStore.save(contractDefinition);
    }
```

However, after requesting the offer catalog, this new field doesn't appear: 

`curl -X POST -H "Content-Type: application/json" -H "X-Api-Key: password" -d @samples/04.0-file-transfer/contractdefinition.json [http://localhost:9192/api/v1/management/catalog/request](http://localhost:9192/api/v1/management/contractnegotiations)`

`{"id":"default","contractOffers":[{"id":"1:a8e4e903-e144-4c86-af1c-1e2c9cc4af62","policy":{"permissions":[{"edctype":"dataspaceconnector:permission","uid":null,"target":"test-document","action":{"type":"USE","includedIn":null,"constraint":null},"assignee":null,"assigner":null,"constraints":[],"duties":[]}],"prohibitions":[],"obligations":[],"extensibleProperties":{},"inheritsFrom":null,"assigner":null,"assignee":null,"target":"test-document","@type":{"@policytype":"set"}},"asset":{"id":"test-document","createdAt":1675675072827,"properties":{"asset:prop:byteSize":null,"asset:prop:id":"test-document","asset:prop:fileName":null}},"provider":"urn:connector:provider","consumer":"urn:connector:consumer","offerStart":null,"offerEnd":null,"contractStart":"2023-02-06T10:17:52.8+01:00","contractEnd":"2023-02-06T10:17:52.8Z"}]}%  
`

How can I add such a field to a contract? 
Thank you for your help, it's really appreciated :)

**Comment by jimmarino** [comment, marked as answer]
Hi, 

In general you probably should not do this but if you are trying to learn, I would look at tutorials on serialzing to JSON using Jackson:

https://github.com/FasterXML/jackson

(You can search for tutorials on how to use Jackson's serialization annotations - there are plenty of them). The correct annotations will solve your problem.
